### PR TITLE
CGSize: Added new `+`, `+=`, `-` and `-=` operator extensions for tuple

### PR DESCRIPTION
**Pull request by hamtiko**
Added new `+`, `+=`, `-` and `-=` operator extensions for scalar.

## Checklist
- [x] I checked the [**Contributing Guidelines**](https://github.com/SwifterSwift/SwifterSwift/blob/master/CONTRIBUTING.md) before creating this request.
- [x] New extensions are written in Swift 5.0.
- [x] New extensions support iOS 10.0+ / tvOS 9.0+ / macOS 10.10+ / watchOS 2.0+, or use `@available` if not.
- [x] I have added tests for new extensions, and they passed.
- [x] All extensions have a **clear** comments explaining their functionality, all parameters and return type in English.
- [x] All extensions are declared as **public**.
- [x] I have added a [changelog](https://github.com/SwifterSwift/SwifterSwift/blob/master/CHANGELOG_GUIDELINES.md) entry describing my changes.

### CODE REVIEW

**Comment by cHaLkdusT** [comment]
Hi all, I think I may disagree with this PR. It is syntactically correct but is semantically incorrect. For consistency in operator overloading, it is best to follow the model's context. 

Just by looking at this code snippet, I need to explicitly look at what is the `result`s type and value. Is it a `CGSize` or a `CGFloat`?
```
let sizeA = CGSize(width: 5, height: 10)
let result = 5 + sizeA
```
Also, in Swift we are not allowed to do infix operation on two different types, they should of the same type.
```
let operand1 = 1
let operand2 = 1.10
let result = operand1 + operand2 // Error: Binary operator '+' cannot be applied to operands of type 'Int' and 'Double'
```

**Comment by hamtiko** [comment]
@cHaLkdusT nice point. But we have operator for multiplying CGSize with scalar (*, *=), do you think these are also incorrect?

As an alternative I think we can have function extensions like this:

```swift
let sizeA = CGSize(width: 5, height: 10)
let result = sizeA.adding(scalar: 5)
```

or even like this:

```swift
let sizeA = CGSize(width: 5, height: 10)
let result = sizeA.resize(dWidth: 5, dHeight: -3)
// result = CGSize(width: 10, height: 7) 
```

**Comment by cHaLkdusT** [comment]
Atleast for me, I think it would be better to deprecate these APIs:
* `static func * (lhs: CGSize, scalar: CGFloat) -> CGSize`
* `static func * (scalar: CGFloat, rhs: CGSize) -> CGSize` 
* `static func *= (lhs: inout CGSize, scalar: CGFloat)`

Or convert these to non-mutating/mutating methods to promote clarity on the API:
* `size.resized(by multiplier: CGFloat)`
* or `size.resize(by mutiplier: CGFloat)`

But if we are just simply mutating either `CGSize`'s `height` and/or `width` and not doing operations on two `CGSize`'s, you just have to declare `var` instead of `let`.

**Comment by guykogus** [comment]
I don't necessarily agree that we can't mix types with infix operators, but I understand the confusion. Especially with something like adding a scalar to a 2D value, (doesn't matter if it were `CGPoint`, `CGSize`, `CGVector`, etc), it's not a mathematically-founded idea, i.e. matrices need to have the same number of rows and columns to be additive. And the reverse, adding a 2D value to a scalar (the example @cHaLkdusT gave in his first comment), does not seem correct to me at all.

However, you can very well multiply 2 matrices so long as the number of columns in the first matrix equals the number of rows in the second (imagine the 2D values to have 1 column and 2 rows).

To me, it makes sense, for example, to multiply `CGSize` by a scalar. You have an image of size 50x50 and you want to double it to 100x100. That's both a regular use case and conceptually obvious. For `CGVector` it also makes perfect sense. `CGPoint` I guess makes sense if you imagine it like a vector, but it's not the same. I'm on the fence on that one, but I think I've actually used it myself, and if you have a use case then that's a good starting point for arguing that it makes sense.

So far I haven't seen any reason for adding or subtracting scalars to `CGSize` that makes more semantic sense than using the operators for using 2 `CGSize` values.

TL;DR I also vote against this inclusion.

@omaralbeik care to weigh in?

**Comment by LucianoPAlmeida** [comment]
> * size.resized(by multiplier: CGFloat)

@cHaLkdusT I really liked the idea, kind of makes it a clear API. But also think that multiply a scalar into a size  using `*` makes sense. 

>  I'm on the fence on that one, but I think I've actually used it myself, and if you have a use case then that's a good starting point for arguing that it makes sense.

@guykogus Maybe an idea that makes more sense is to have an operator that applies on tuples. 
The major gain I've seen in this is that you can make operation one-liners. e.g.
```swift
var size = CGSize(width: 10, height: 10)
size.width += 5
size.height += 5

size += 5 // Although it may seem not right it simplify syntax
```
So my proposal is to make it operate on tuples:
```swift
// We can make one liners
size += (width: 5, hight: 5) // That could simplify things and make more sense semantically than adding to a  scalar.
//or
size += (5, 5) 
```

**Comment by guykogus** [comment]
@LucianoPAlmeida I like the idea of the tuple. It keeps it semantically correct and simplifies the syntax, which I believe was the original idea. How does that sound to you, @hamtiko?

**Comment by hamtiko** [comment]
@guykogus @LucianoPAlmeida I like the idea too.

Will do updates today.

Thanks.

File: Sources/SwifterSwift/CoreGraphics/CGSizeExtensions.swift [hunk 1]
```diff
@@ -82,9 +82,37 @@ public extension CGSize {
...
     ///
     ///     let sizeA = CGSize(width: 5, height: 10)
+    ///     let result = (4, 5) + sizeA
+    ///     // result = CGSize(width: 9, height: 15)
+    ///
+    /// - Parameters:
+    ///   - scalar: tuple value.
+    ///   - rhs: CGSize to add to.
+    /// - Returns: The result comes from the addition of the given tuple and CGSize.
+    static func + (tuple: (width: CGFloat, height: CGFloat), rhs: CGSize) -> CGSize {
```
Comment (guykogus) [inline Sources/SwifterSwift/CoreGraphics/CGSizeExtensions.swift:109]: As mentioned in the discussion, this version of the operator doesn't make sense. Imagine that with `+` you should be able to have an equivalent assignment operator (`+=`). In the assignment operator for a `CGSize` to the tuple, e.g.
```Swift
var size = (CGFloat(), CGFloat())
size += CGSize.zero
```
`size` would still be a tuple, not a `CGSize`. Therefore it doesn't make sense if the equivalent `+` operator returns a `CGSize`.
I recommend that you delete this method, don't just change the return type.

Comment (hamtiko) [reply Sources/SwifterSwift/CoreGraphics/CGSizeExtensions.swift:109]: Ok, I see. Will remove it.
Thanks.

File: Tests/CoreGraphicsTests/CGSizeExtensionsTests.swift [hunk 2]
```diff
@@ -62,6 +62,13 @@ final class CGSizeExtensionsTests: XCTestCase {
         XCTAssertEqual(result.height, 14)
     }
 
+    func testAddTupleRight() {
```
Comment (guykogus) [inline Tests/CoreGraphicsTests/CGSizeExtensionsTests.swift:65]:
```suggestion
    func testAddTuple() {
```

File: Tests/CoreGraphicsTests/CGSizeExtensionsTests.swift [hunk 3]
```diff
@@ -86,6 +107,13 @@ final class CGSizeExtensionsTests: XCTestCase {
         XCTAssertEqual(sizeA.height, 6)
     }
 
+    func testSubtractTupleEqual() {
```
Comment (guykogus) [inline Tests/CoreGraphicsTests/CGSizeExtensionsTests.swift:110]:
```suggestion
    func testSubtractEqualTuple() {
```

File: Tests/CoreGraphicsTests/CGSizeExtensionsTests.swift [hunk 4]
```diff
@@ -62,6 +62,13 @@ final class CGSizeExtensionsTests: XCTestCase {
         XCTAssertEqual(result.height, 14)
     }
 
+    func testAddTupleRight() {
+        let sizeA = CGSize(width: 5, height: 10)
```
Comment (guykogus) [inline Tests/CoreGraphicsTests/CGSizeExtensionsTests.swift:66]:
```suggestion
        let size = CGSize(width: 5, height: 10)
```

File: Tests/CoreGraphicsTests/CGSizeExtensionsTests.swift [hunk 5]
```diff
@@ -70,6 +77,13 @@ final class CGSizeExtensionsTests: XCTestCase {
         XCTAssertEqual(sizeA.height, 14)
     }
 
+    func testAddEqualTuple() {
+        var sizeA = CGSize(width: 5, height: 10)
```
Comment (guykogus) [inline Tests/CoreGraphicsTests/CGSizeExtensionsTests.swift:81]:
```suggestion
        var size = CGSize(width: 5, height: 10)
```

File: Tests/CoreGraphicsTests/CGSizeExtensionsTests.swift [hunk 6]
```diff
@@ -78,6 +92,13 @@ final class CGSizeExtensionsTests: XCTestCase {
         XCTAssertEqual(result.height, 6)
     }
 
+    func testSubtractTuple() {
+        let sizeA = CGSize(width: 5, height: 10)
```
Comment (guykogus) [inline Tests/CoreGraphicsTests/CGSizeExtensionsTests.swift:96]:
```suggestion
        let size = CGSize(width: 5, height: 10)
```

Comment (hamtiko) [reply Tests/CoreGraphicsTests/CGSizeExtensionsTests.swift]: Nice catch.
Done.
Thanks.
